### ISSUE TITLE
Version 4.10717 breaks Blyss protocol

### ISSUE THREAD

**Issue by Jvsjvs** [domoticz/domoticz#3232]
The Blyss protocol is broken in the update. 

When I look at the log files I see that Domoticz sends the wrong command string. 
It sends: RFLink Sending: **10;Blyss;000000;?�;ON;** 
This should be:RFLink Sending: **10;Blyss;003FCC;A1;ON** 
(003FCC is my address)
This results in a time out.

In the file: "**RFLinkBase.cpp** " in **lines 341 to 346** it incorrectly builds the command string to send. I personally adapted these lines (removed the different handling in case of Blyss) and did a rebuild. After that the Blyss protocol is working normally again.

Can somebody fix these 5 lines so that I can update to a Beta release again instead of having to use my own build?

Thanks,

John.

**Comment by gizmocuz** [comment]
Could you try replacing the line with 'sstr' with

`			sstr << switchtype << ";" << std::hex << std::nouppercase << std::setw(6) << std::setfill('0') << (pSwitch->id >> 16) << ";" << szUnit << ";" << switchcmnd << ";";
`

**Comment by Jvsjvs** [comment]
Hi,

Thanks for looking into this.

Unfortunately that does not fix it. The result is the same, the address still equals "000000" and the "button" string szUnit is also incorrect, exactly as indicated above.

If I just use the "sstr line" of the old code of the previous release everything works fine (remove the special case for Blyss) :-).

So both the szUnit string as the switch_id are incorrectly constructed.
Using "pSwitch->id" instead of "pSwitch->id >> 16" (removing the shift) fixes the address.
Using "pSwitch->unitcode" instead of the szUnit fixes the unit id (removing the shift) fixes the unit id.

What are you trying to achieve with shifting both parameters??

Thanks,

John.

**Comment by gizmocuz** [comment]
No idea, someone suggested to do so...

**Comment by gizmocuz** [comment]
So you remarked the whole block ? so it's just

```
		sstr << "10;";
		sstr << switchtype << ";" << std::hex << std::nouppercase << std::setw(6) << std::setfill('0') << pSwitch->id << ";" << std::hex << std::nouppercase << pSwitch->unitcode << ";" << switchcmnd << ";";

```

**Comment by Jvsjvs** [comment]
Correct, then it works like a charm.

This was also the way it was in the previous release (identical to any of the other RFLink protocols).

Can you change it like this in a beta release such that I can again remove my own code?

Thanks,

John.

**Comment by gizmocuz** [comment]
Sure, done in beta 10796

**Comment by gbuico** [comment]
@gizmocuz 
Hi gizmocuz.
I'm in trouble because all my Blyss devices are not controllable anymore, but I'd love to stay on a stable version.
Is there any way to apply your fix (build 10796) while remaining on stable 10717 ?
May be this is a stupid question but I've no clue on how the build is compiled
Thanks
Giacomo

**Comment by gizmocuz** [comment]
@gbuico , no thats not possible, but once the zwave issues have been resolved there will be a new stable soon
If you do not use zwave, just update to the beta

**Comment by gbuico** [comment]
Great. not using zwave so I'll do it immediatly.
Thanks a lot
Giacomo

**Comment by 120fr** [comment]
Hi,

There is still a problem Blyss
Current build 4.10817

Correct RFlink send
10;Blyss;002cac;A4;UP

after update
10;Blyss;2cac4134;A4;UP

**Comment by gizmocuz** [comment]
@120fr  that means the previous code was correct?
I reverted it again... maybe you two have different issues
Anyway, i can not test this, best to debug it and make a PR if possible ....

Have a look at the commit changes what and where it's changed

https://github.com/domoticz/domoticz/commit/636d15f6c0f631aa535b427ee54ef36ec7176197

@120fr , what's the current string being send with beta 10822 is this correct?

**Comment by 120fr** [comment]
beta 10822

RFLink Sending: 10;Blyss;002cac;11;UP;

Well done

**Comment by gbuico** [comment]
@gizmocuz 
I'm currently restoring an SD image of v9700 where Blyss was working, note done what was coming out, then upgrade to 10717 and nothing down the wrong one... 
After all this I'll install the last beta and do a third check.... and let you know.
If you have special indication about where/how install a specific beta let me know... 
Giacomo

**Comment by Jvsjvs** [comment]
@Gizmocus,

Reverting this is a mistake!

**Now all the normal Blyss devices no longer work**. 

The support for the Blyss devices is already working for ages using the old code before the 4.10717 release (I am already using it for years). Making the exception in lines 342 to 347 in the RFLinkbase.cpp file does not make any sense and breaks with all the users that have been using the Blyss devices for years. 

If you look at the RFLink protocol specification, then you see that **there is no specific case for the Blyss protocol**. you can find the RFLink protocol description here: http://www.rflink.nl/blog2/protref. So why would we for this protocol suddenly shift out the data???. Please point at the spec point that would justify this shift.

@12fr, most likely you have configured your address wrong when installing the Blyss devices in Domoticz.

@Gizmocus, please change back to the way it always has been, the way it always has been working and the way it is according to the specification of the protocol on the RFLink page referenced above.

Thanks,

John.

**Comment by gizmocuz** [comment]
@Jvsjvs , you are causing confusion here ! First, in there is no version 4.07107, you probably mean 10717, but maybe you mean it was working in 4.9700 ?

@12fr, how did you add the Blyss to Domoticz?
It could indeed that you added it wrong...

I have reverted the patch again... @12fr, i suspect when you add it correctly it should indeed work as in your case the unitcode is part of the ID, and @jvsjvz has it separate (as it should)

**Comment by gbuico** [comment]
@gizmocuz 
I can confirm what you say is valid for me. Blyss were working fine on 4.9700 and stopped working on 10717.
Unfortunately I'm having issues with my disk images, but I'll come back with more info as soon as possible

**Comment by 120fr** [comment]
@gizmocuz

My blyss switches have been added with the remote control blyss
https://www.castorama.fr/telecommande-programmable-lcd-blyss/3454975828836_CAFR.prd#

I first initialize the switch with the remote control, then I registered the command in automatic detection in domoticz.

**Comment by gbuico** [comment]
@gizmocuz 
So, here are the results.
Restored the SD image of Domoticz 10717 plus all Raspbian updates. Blyss non working
In the log I see a line  (like @Jvsjvs  ) sayng "Status: RFLink Sending: 10;Blyss;000000;?t;DOWN;

Allowed Beta version and done the upgrade to Domoticz 10826... Blyss working like a charme.
The only question mark for me at the moment is that the log at level 10826 is not showing the above "Status" line with the actual command sent to RFLink.
Weird because I just upgraded and not changed any setting.
Will investigate further but right now my Blyss are back in service

Of course... I'm available for whatever additional test you may need
Giacomo

**Comment by gbuico** [comment]
@120fr 
If I can ask, which remote are you using and what are you trying to control?.
Here I'm using the remote 582883 (Castorama) to control the shutters (external module not the one integrated in the shutter)

**Comment by Jvsjvs** [comment]
@gizmocuz, sorry I indeed meant: "before the 4.10717 release", before this release it was working fine. I edited my comment.

Thanks for reverting it back to the old release, now it works again.

**Comment by gbuico** [comment]
@gizmocuz 
I have good/bad news on this issue… but at least something is getting clear and both @Jvsjvs and @120fr may be right in their environment.
As I said before I had my Blyss working fine in 4.9700. Upgrading to 10717 they stopped working.
You removed the extra lines of code in 10826, I upgraded and Blyss were working again.
Of course I did not acquire again the remote key. I just carried out what was working in 4.9700.

Today I wanted to reinstall Domoticz from scratch on a new RPi, so I installed Domoticz beta 10826, connected the RFLink and loaded again in Domoticz the key of my remote.
Everything went fine BUT... the devices newly created in Domoticz have no effect at all..meaning.. shutters not moving even though the Domoticz Log show that the command was sent.

Summarizing. RFlink remote commands acquired in 4.9700 are not working in 10717 and working in 10826. Remote commands acquired in 10826 don't work on 10826.

Now, this effect can be seen also if you look at the same remote action in the Device panel of Domoticz.
This picture shows how the Blyss devices are when they are defined in 4.9700 and used in 10826.
All five ID's are identical and are 00001674
![old def in beta](https://user-images.githubusercontent.com/34986440/58385719-0343c980-7ff5-11e9-9f6f-a2ca3f40139a.jpg)
Now. look at how they are when acquired in 10826
![new def beta](https://user-images.githubusercontent.com/34986440/58385762-a72d7500-7ff5-11e9-8e96-883a798ec4ee.jpg)
Now the id has the famous 1674 on the left and each button as different four numbers on the right
Unit number is unchanged.

Bottom line, not knowing who, why and when somebody asked you to change this piece of code. my guess is that in 10826 you removed the new code related to the usage of the remote command... but somewhere you still have some new coding acquiring the remote key press in a different way.
I hope I've been clear enough... if not please ask.
In order to furtehr contribute I have another issue but will write it in another post

Giacomo

**Comment by gbuico** [comment]
Now... really in short. 
In the past and up to 10717 included, the log of the RFLink command was like this:
![log up to 10717](https://user-images.githubusercontent.com/34986440/58385830-c2e54b00-7ff6-11e9-914b-a883b553441c.jpg)

As you see, there is a Status line showing which command has been actually sent to RFLink
In 10826 the Log became like this... no trace of the Status line.. so.. no way for me to contrinute further
![log in beta](https://user-images.githubusercontent.com/34986440/58385851-19528980-7ff7-11e9-8905-fab2ad1c3ba5.jpg)

Is there any easy way to go back to the previous logging?
Now... what I would really like to do now is to do the following.
Go back to 10717 and try to acquire and use the remote keys.
I'd bet that it might work... but I have no overview on the added code so I will wait your comments before making any change to my setup.
So.. be aware that I have a Rpi with old definitions in beta (working) and an Rpi with new definitions in the same beta (not working)... and redy to change/test whatever you will want

Giacomo

**Comment by gbuico** [comment]
@gizmocuz  URGENT URGENT
PLEASE READ THIS ONE FIRST !!

After all the test I did yesterday (described above) I couldn't resist to verify my theory.
I found a spare SD card, installed 10717 and acquired from scratch the keypress of the remote to be used with RFlink.
Then I tried to use the newly acquired devices and THEY WORK !

This means: the whole new code for Blyss in 10717 works perfectly but all the devices (remote keypress) must be re acquired using 10717.
Devices acquired with earlier versions (i.e. 4.9700) will NOT work.

Bottom line, I have no clue why the new code has been added in 10717 but, assuming that there was a reason, my opinion is that the code for this device should be the one currently in 10717, adding a reminder to reacquire the RFlink devices.

Leaving the code at level 10826 would mean support the old devices but being unable to add new.
Ready to provide additional info if you need
Giacomo

@Jvsjvs @120fr , FYI

**Comment by Jvsjvs** [comment]
So no clue why the code has changed and …
the protocol is not different for Blyss then any of the other protocols for the RFLink and ...
the 4.9700 solution works fine if you register and use your devices with that solution.


The changed code does not work for me, which means I have to somehow fix 49 devices :-(


So let's keep it according the protocol specification, then it works.

**Comment by gbuico** [comment]
@Jvsjvs 
Personally I have no clue why the code is changed but what I see now is that each device has a unique ID where the leftmost 4 digit are the remote ID and the four right are the pressed button.
I would assume that @gizmocuz did not change it for fun... but at least now we know that we either stay with the old way of acquiring and old way of making use of RFlink (ref. 4.9700) or we stay with the new version (10717). Beta 10826 is an unusable mess allowing the existing devices to be used and no way to define new ones

**Comment by gizmocuz** [comment]
@gbuico , i was asked to change it a while ago. (I do not use RFLink myself with Blyss)
The current code in the beta (and github) is the same code as 4.9700
I did not change the firmware of the RFLink of course.. what i do understand is that there are new Blyss devices
It should however be possible to detect this.

Are the Blyss devices automatically detected, or do you have to manually add them in Domoticz ?
If you have to manually add them, then i suspect here is the issue

**Comment by gbuico** [comment]
@gizmocuz 
Let me answer with order.
Blyss devices have to be added manually.
You press a key on the remote and it appears as a Device.
Of course, if you already have that Device, there is no need to add it when you change Domoticz version.

I don't remember when I added my Blyss devices but must have been before 4.9700

What I can tell you for sure is that Blyss stopped working when I migrated to 10717 BUT deleting and adding them back solves the problem and the definition of the device looks more accurate.

What I can tell you for sure is that if you carry the existing Blyss devices to 10826 they keep working without doing anything BUT if you add something new (still Blyss) it won't work.

My conclusion is that the best code we have right now is the one in 10717 because it's the only one allowing you to add new Blyss devices with the small minus that you need to add back the old ones.

One last remark. Yesterday I added with success in 19717 all the keypress of the most complete Blyss remote available at the moment (model 5828839) and everything is working fine... 


Bottom line... 10717 is the real stable... you just need to delete and add again the existing Blyss devices

**Comment by gizmocuz** [comment]
@gbuico , thats automatic (You press a key on the remote and it appears as a Device)
Manual is from the Switches tab, press the 'Add Manual' button, specify the hardware, select the type, address/unit code etc
That implies when you where using 10717 (with the patched code), you should just have received new devices (when you have accept new hardware/devices enabled in the settings)

**Comment by gbuico** [comment]
@gizmocuz sorry. I misunderstood you.
No need for Manual add.
And yes, in 10717 you just need to delete the old Blyss and add them back
As you can see from the log the protocol is just perfect and in line with the specs
![10717 final](https://user-images.githubusercontent.com/34986440/58458383-5e76d880-8129-11e9-86ce-7d249a6ea955.jpg)

Devices carried directly from 4.9700 were producing this wrong call:
RFLink Sending: 10;Blyss;000000;?�;ON;

**Comment by gizmocuz** [comment]
Beta 10848 has the new code once again. Hopefully all is clear and solved now and we can close this issue ?

**Comment by gbuico** [comment]
Fine for me @gizmocuz . Thank you

**Comment by gizmocuz** [comment]
@Jvsjvs , works for you as well ?

**Comment by Jvsjvs** [comment]
Hi Gizmocuz,

Thanks for asking. Unfortunately the answer is no :-)

It does not work for me as I have the pSwitch->id contain the actual id that is not to be shifted, and there is no keypress data appended that has to be removed. So this fix kills my id (becomes 0000). The pSwitch->unitcode is also shifted to 00 for me.

I can of course change my 49 device IDs and shift them to the left such that they can be shifted to the right again in the code to make it work :-)

I think that this change is wrong though. If you now look at the code then ID does no longer contains the ID (only for Blyss devices), but a shifted version with keypress data appended. Apparently the data is not properly captured when recording those Blyss devices. So it would be better to change the code that automatically captures the Blyss devices (which I don't use as I add them manually). Then the code is at least clean and it works for all of us.

I also still don't understand (did not look into it) why all the Blyss devices did work in the versions upto 4.10714 out of the box with the old code. So still I don't see wy this change is done breaking existing users their systems?? Also gbuico confirms that with the old code his devices worked properly.

Is it possible to revert this and look at the capturing code to fix this?

Thanks,

John.

**Comment by gizmocuz** [comment]
Hi John

Sorry about that... But I am in the middle of two fixes here.
So please check the blyss protocol, contact the RFLink author ... 
I am not touching this any longer... This patch was in for a long long time now

**Comment by Jvsjvs** [comment]
Which version did it get in then?
It was not part of the previous stable release. As soon as I updated to the new stable release it was broken.

I understand that the different views make it complex.

In my view it is not good to break with existing devices, or making the change without confirmation that it works with existing users.

**Comment by gbuico** [comment]
@Jvsjvs 
While I'm saying that with the old code the devices worked properly (4.9700), I'm also saying that removing it for some reason prevent adding new devices.
My existing devices were coming from a version before 4.9700 so I don't know how and when the bug came in.
What I know for sure is that with 10717 you are able to add from scratch each device and it works properly once acquired.

Having tried all the combinations, in my opinion right now 10717 is the only working version allowing to add and operate the Blyss devices... the other versions either invalidate old or new devices

**Comment by gizmocuz** [comment]
@gbuico , version 10717 is the same as the current 10850+ version
@Jvsjvs , i agree, it is breaking old switches... remember it was not me that proposed (this needed apparently) change, I don't own Blyss and do not use RFLink for switches
But good to know this is working now, best to look at the Blyss protocol (maybe from the RTL433 project) to see how the actual ID is build up see if it is correct now.
For now, closing the issue

**Comment by Oeil2L** [comment]
I made the mistake to update domoticz and my blyss stop working as explain by others.

Solution based on this link:
https://easydomoticz.com/forum/viewtopic.php?t=4054

I put on a script .sh this command:
echo "10;Blyss;ff98;A1;OFF;" > /dev/ttyUSB0

and call this script from domoticz, same for ON action.

**Comment by gizmocuz** [comment]
@Oeil2L , this has been changed a long long time ago. And there are blyss users using the latest domoticz version.
If you enable 'accept new hardware/devices' in the settings, do you get new switches ? If yes, can you use those ?
Did you try finding the answer here ?

https://github.com/domoticz/domoticz/issues?utf8=%E2%9C%93&q=blyss

**Comment by Jvsjvs** [comment]
Although this has been changed long ago, it is incorrect and people (like me) have let go of Blyss and Domiticz since it does not work.

**Comment by gizmocuz** [comment]
@Jvsjvs , the problem is that it was requested to change it (seems to be a newer version issue), then change it back, then change it back again, and back again...
I don't own Blyss hardware, so I am unable to test.
Maybe once i have a old and new type i could try to make it compatible.

For now, best to ask on the forum what we should do and submit a pull request with the needed changes

**Comment by Oeil2L** [comment]
@gizmocuz 
Thanks for your answer.
To be clear, I blame nobody for this changes, probably it was necessary (my devices are pretty old)
I share only a solution.

After the update, my power switch stopped working, I made a new automatic detection, a new device was created, I modified all my scenes/groups and everything went well.
My previous device on Domoticz was never detected again as if its internal code had been changed.

But (as always;)) I create a manual blyss device to pair it with power socket because the previous one doesn't work anymore, nothing, I tried a lot of ID/Type code, no reaction.
I sent the previous ID code by script directly to Rflink, it worked perfectly.

Maybe the Blyss protocol has evolved, or my old device didn't respect something, or the version firmware of my Rflink is outdated.
